### ISSUE TITLE
.Distinct() does not work on Db.SQL<Starcounter.Query.Execution.Row>() results.

### ISSUE THREAD

**Issue by eriksunsol** [Starcounter/Starcounter#3682]
(Starcounter version 2.2.1.2180)

When running a query that returns multi-value rows, e.g. (conceptually):

```C#
    var result = Db.SQL<Starcounter.Query.Execution.Row>(
                    "SELECT m.Source, m.Destination FROM MyProj.Mediator");
```
and wanting only the unique combinations of `m.Source` and `m.Destination` by doing:

```C#
    var distinct = result.Distinct();
```

you end up with `distinct` having the same content as `result` even if there are duplicate entries.

To get correct behavior for this case, an `IEqualityComparer` is needed:

```C#
    public class Row2Comparer : IEqualityComparer<Starcounter.Query.Execution.Row>
    {
        public bool Equals(Row x, Row y)
        {
            for (int i = 0; i < 2; ++i)
                if (x.GetObject(i) != y.GetObject(i)) 
                    return false;

            return true;
        }

        public int GetHashCode(Row obj)
        {
            return obj.GetObject(0).GetHashCode() + obj.GetObject(1).GetHashCode();
        }
    }
```

Using `Row2Comparer` like this works as expected:

```C#
    var distinct = result.Distinct(new Row2Comparer());
```

This is an ugly workaround though. The proper solution would be for `Starcounter.Query.Execution.Row` to implement proper comparison support, e.g. proper overrides of `object.Equals()` and `object.GetHashCode()`and/or implement `System.IEquatable<Starcounter.Query.Execution.Row>`.

**Comment by miyconst** [comment]
@eriksunsol thank you for the suggestion.

Please be aware that using `Starcounter.Query.Execution.Row` is not recommended and your solution can be easily solved with a bit of LINQ.

```cs
var result = Db.SQL<MyProj.Mediator>("SELECT m FROM MyProj.Mediator m").Select(x => new { x.Source, x.Destination }).Distinct();
```

cc @k-rus

**Comment by eriksunsol** [comment]
@miyconst that's actually quite elegant. I'll switch to using LINQ to produce the pair.

Where is it documented that using `Starcounter.Query.Execution.Row` is not recommended? Actually I can't find any documentation at all on `Starcounter.Query.Execution.Row`, so maybe that's an indication... :-)

Anyway, the LINQ solution looks is nicer, so I wouldn't mind if `Starcounter.Query.Execution.Row` is marked as `[Obsolete]` and the support eventually dropped entirely.

**Comment by miyconst** [comment]
`Starcounter.Query.Execution.Row` is not `[Obsolete]`. It is used when you execute a query and don't have reference to the types returned. For example in Administrator you should be able to query all classes from all apps while Administrator does not have a reference to those classes.
Also it is useful for debugging purposes.

>Where is it documented that using `Starcounter.Query.Execution.Row` is not recommended? Actually I can't find any documentation at all on `Starcounter.Query.Execution.Row`, so maybe that's an indication... :-)

Exactly, you are trying to use a non-documented feature.

**Comment by miyconst** [comment]
@eriksunsol I have created another issue to document the `Starcounter.Query.Execution.Row` class - https://github.com/Starcounter/Starcounter/issues/3683. Let's close this one.

**Comment by eriksunsol** [comment]
@miyconst Well, doesn't it still make sense to have `Starcounter.Query.Execution.Row` behave as one would expect when using comparisons, e.g. for use with `.Distinct()`? Sure, there is a workaround (which I actually prefer) and the documentation should explain proper use, but I don't see how that is an excuse for closing this issue.

**Comment by miyconst** [comment]
Let's see what @k-rus says.

**Comment by k-rus** [comment]
Thank you for the question. I am not a right person to ask about .NET interfaces on top of SQL. May be @per-samuelsson has better knowledge.

From SQL point-of-view `DISTINCT`needs to be supported natively in SQL and it will happen in future.

 I don't know if supporting C# method `Distinct()` by implementing `IEqualityComparer` will be done for the current version of code host. May be @dan31 or @per-samuelsson can clarify on this.

I think @miyconst suggestion might be good enough. This depends on default implementation of LINQ. Enumerator semantics of executing simple SQL queries corresponds `IEnumerator`. So if the above example enumerates only once and materialises into temporal collection, then the semantics will be the same. On the other side if the materialisation allocates objects in the heap, then it can bring unexpected slowness to the system.

**Comment by eriksunsol** [comment]
I was just looking at this again (was going to re-factor my `.Distinct()` LINQ query a bit) and realized this is related to #1694 and #3689, because my use of `.Distinct()` where `m.Source` and `m.Destination` as exemplified above are also database classes.

Whatever the equality comparer used by `.Distinct()`, be it the default comparer or the one provided by the user code, there is an element of doubt as to whether this will work as expected due to the questions discussed in #1694. I assume in this case since the objects come from a single query, comparison will work as expected, but from a user code programmer's perspective this is an unknown.

If Starcounter provides an implementation of equality comparison on `Starcounter.Query.Execution.Row` and takes #1694 into account however appropriately, then at least this element of doubt can be eliminated.

**Comment by k-rus** [comment]
> x.GetObject(i) != y.GetObject(i)

Just a side note that this actually will not work as expected due to #1694. It should be:

```c#
x.GetObject(i).Equals(y.GetObject(i))
```

**Comment by eriksunsol** [comment]
@k-rus 

(missing negation, should be: `!x.GetObject(i).Equals(y.GetObject(i))`, but that's minor...)

Actually `x.GetObject(i) != y.GetObject(i)` worked just fine when I tried it (probably because the instances came from the same query), but as discussed in #1694 this may not work in other cases.

**Comment by per-samuelsson** [comment]
@eriksunsol 

Thanks for taking the time providing us with all this feedback. It's certainly appreciated!

Just to make this clear here: `Distinct` is a `Linq` extension method, operating on collections, and I'm not surprised at all it use standard comparisons for the platform being (.NET). And in such context, equality between reference types is conducted on addresses, not by-value.

Hence, we could translate this issue to a standard console application snippet like this:

```cs
class Bar {
    public string Value;
}

class Program {
    static void Main() {
        var b1 = new Bar() { Value = "Foo" };
        var b2 = new Bar() { Value = "Foo" };

        var bars = new[] { b1, b2 };

        foreach (var b in bars.Distinct()) {
            Console.WriteLine(b.Value);
        }
    }
}
```

This would certainly produce

> "Foo"
> "Foo"

I'm sure you know all this. I just want to take the chance to get it documented.

Now, is that *correct* behavior? Well, it's hard to imagine having the runtime doing something else as a default, so in a sense it is. And hence it's up to the developer of the `Bar` class to improve it as he think fits (and, more important, as he think will be of least surprise to the target audience).

Could we improve our `Row` class accordingly? I guess we could. Without knowing anything about it's current implementation, I'd guess it would be possible to iterate all values of such instance, invoking the most appropriate comparisons for each data type (which, in our case, probably would be `Equals` for reference types, and standard comparison for all the rest, just as the platform would).

As always, it's also a question on resources and how to prioritize, and I'm not sure how this one ranks. Especially in the context of the fact that we have a future version coming up, implementing native SQL DISTINCT, as @k-rus point out.

Finally, one idea could be to implement just a rather trivial, but *optional* built in `IEqualityComparer<Starcounter.Query.Execution.Row>`, similar to the one you provided in the OP. But it would need to take additional things into account o/c, comparing the full cargo of in-going rows, not just reference fields.
